### PR TITLE
Supporting ReScan of HashJoin with Spilled HashTable

### PR DESCRIPTION
**Pull request by foyzur**
To support RecursiveCTE we need to be able to ReScan a HashJoin as many times as the recursion depth. The HashJoin was previously ReScannable only if it has one memory-resident batch. Now, we support ReScannability for more than one batch. The approach that we took is to keep the inner batch files around for more than the duration of a single iteration of join if we detect that we need to reuse the batch files for rescanning. This can also improve the performance of the subplan as we no longer need to materialize and rebuild the hash table. Rather, we can just reload the batches from their corresponding batch files.

To accomplish reloading of inner batch files, we keep the inner batch files around even if the outer is joined as we wait for the reuse in subsequent rescan (if rescannability is desired).

The corresponding mail thread is here: https://groups.google.com/a/greenplum.org/forum/#!searchin/gpdb-dev/Rescannability$20of$20HashJoin%7Csort:relevance/gpdb-dev/E5kYU0FwJLg/Cqcxx0fOCQAJ

### CODE REVIEW

**Comment by hlinnaka** [comment]
To be clear, this is just a performance improvement, right? Hash joins have always been rescannable, it's just that it might need to redo the whole join, rescanning both child nodes. While with this patch, it only needs to rescan the outer side.

I would prefer to keep the code as unchanged from upstream as possible. This is a nice performance improvement, but can we get that done in the upstream first, and get it into GPDB later, please.

**Comment by foyzur** [comment]
Hi @hlinnaka thanks for looking into the PR! Although this is for performance, the alternative solution is lot more expensive with recursive CTE. 

The problem of GPDB hash join is it only supports rescan over materialize (or when it has only one batch, resident in the memory) or a subtree that it can rescan and rebuild the hash table every time. Unfortunately, we may have hash join over motion that we need to rescan. With recursive CTE this became a more prominent requirement as we now deliberately place WorkTable on the other side that keeps changing and requires rescanning the inner side. Throwing out the hashtable and rescanning the subtree could be very expensive as we need a materialize (the motion does not support rescanning) and rebuild hash table every time.

**Comment by foyzur** [comment]
@vraghavan78 what can we do about postgres pushing first?

File: src/backend/executor/execWorkfile.c [hunk 1]
```diff
@@ -358,15 +358,22 @@ ExecWorkFile_Rewind(ExecWorkFile *workfile)
...
-			file_size = bfz_append_end((bfz_t *)workfile->file);
-			ExecWorkFile_AdjustBFZSize(workfile, file_size);
+			f = ((bfz_t *)workfile->file);
+			if (f->mode == BFZ_MODE_APPEND)
+			{
+				file_size = bfz_append_end((bfz_t *)workfile->file);
+				ExecWorkFile_AdjustBFZSize(workfile, file_size);
+			}
+
+			f->mode = BFZ_MODE_FREED;
```
Comment (kavinderd) [inline src/backend/executor/execWorkfile.c]: A comment about what's happening in the BFZ case would be helpful

File: src/include/nodes/execnodes.h [hunk 2]
```diff
@@ -2274,6 +2274,7 @@ typedef struct HashJoinState
 
 	/* set if the operator created workfiles */
 	bool workfiles_created;
+	bool rescannable; /* Do we need to support rescan for this hash join */
```
Comment (hsyuan) [inline src/include/nodes/execnodes.h]: Hmm, after a second thought, I think we'd better give it some name like `hashtable_reusable` or `reuse_hashtable`, meaning the hash table of this hash join is reusable or not, if not, the hash table has to be rebuilt every time we rescan the hash join. I have to admit that `rescannable` is a little misleading.

**Comment by hlinnaka** [comment]
On 07/14/2017 12:05 AM, foyzur wrote:
> Hi @hlinnaka thanks for looking into the PR! Although this is for
> performance, the alternative solution is lot more expensive with
> recursive CTE.
>
> The problem of GPDB hash join is it only supports rescan over
> materialize (or when it has only one batch, resident in the memory)
> or a subtree that it can rescan and rebuild the hash table every
> time. Unfortunately, we may have hash join over motion that we need
> to rescan. With recursive CTE this became a more prominent
> requirement as we now deliberately place WorkTable on the other side
> that keeps changing and requires rescanning the inner side. Throwing
> out the hashtable and rescanning the subtree could be very expensive
> as we need a materialize (the motion does not support rescanning) and
> rebuild hash table every time.

Ah, looking at the code more closely now, I see that you have currently 
completely disabled hash joins with recursive CTE plans, and this patch 
re-enables it. And the reason is that it would require re-scanning the 
hash join, which would outright fail:

postgres=# with recursive cte (level) as (
select 1
union all
select level+1 from cte, bar where level = c)
select * from cte ;
ERROR:  Illegal rescan of motion node: invalid plan. (nodeMotion.c:1557)
HINT:  likely caused by bad NL-join, try setting enable_nestloop off

Ok, I agree we need to do something about that. A quick solution might 
be to add a materialize node to the inner side. Not sure if that's easy 
to do in the planner.

- Heikki

File: src/backend/executor/nodeHashjoin.c [hunk 3]
```diff
@@ -1115,8 +1086,9 @@ ExecReScanHashJoin(HashJoinState *node, ExprContext *exprCtxt)
 	 */
```
Comment (hlinnaka) [inline src/backend/executor/nodeHashjoin.c]: The comment just above this is out-of-date.

File: src/backend/executor/nodeHashjoin.c [hunk 4]
```diff
@@ -1289,4 +1269,107 @@ isHashtableEmpty(HashJoinTable hashtable)
 	return isEmpty;
 }
 
+/*
+ * In our hybrid hash join we either spill when we increase number of batches
+ * or when we re-spill. However, the first batch is always memory resident.
+ * This optimization requires us to save the first batch separately if we
+ * need to support rescanning (e.g., recursive CTE).
+ */
```
Comment (hlinnaka) [inline src/backend/executor/nodeHashjoin.c]: I don't understand this. If the first batch is always kept in memory, why do we need to save it to disk and reload?

**Comment by hsyuan** [comment]
If the hash table has multiple batches, the 1st batch will be discarded
when we load later batches into memory. Next time if we rescan the
hashjoin, we have to reload the inner tuples of 1st batch into hashtable,
otherwise, the outer tuple of 1st batch will match nothing. Does this make
sense?

~ ~ ~
Haisheng Yuan

**Comment by hsyuan** [comment]
Yes, add a materialize node on top of Motion node would be a quick
solution, just like how subplan does. It should be easy in the planner,
because subplan already does that.

~ ~ ~
Haisheng Yuan

File: src/backend/executor/nodeHash.c [hunk 5]
```diff
@@ -989,7 +992,7 @@ ExecHashTableInsert(HashState *hashState, HashJoinTable hashtable,
 			}
 		}
 	}
-	else
+	else if (hashtable->first_pass)
```
Comment (foyzur) [inline src/backend/executor/nodeHash.c]: The batch file can still contain tuples from another batch. If, e.g., we started with two batches and spilled few tuples, the mod function may put some tuple in batch 2 (e.g.). Later on we found out that we need to double the batch. If we then have 4 batches, we only spill future (and the then in memory tuples) to appropriate batches, considering the full set of 4 batches (i.e., mod will be 4). But, later on, when we load the spill file, some tuples in batch 2 (before doubling the number of batches) will actually fall in other batch (as the number of batches has increased).

File: src/backend/executor/nodeHashjoin.c [hunk 6]
```diff
@@ -1133,8 +1105,16 @@ ExecReScanHashJoin(HashJoinState *node, ExprContext *exprCtxt)
...
-			node->hj_HashTable->curbatch = 0;
+			if (node->hj_HashTable->nbatch > 1)
+			{
+				/* Force reloading batch 0 upon next ExecHashJoin */
+				node->hj_HashTable->curbatch = -1;
+			}
+			else
+			{
+				/* MPP-1600: reset the batch number */
+				node->hj_HashTable->curbatch = 0;
```
Comment (foyzur) [inline src/backend/executor/nodeHashjoin.c]: If we have only one batch that will always be memory resident. So, nothing to reload. Just set the curbatch to 0.

File: src/backend/executor/nodeHashjoin.c (hunk 4 above)
Comment (foyzur) [inline src/backend/executor/nodeHashjoin.c]: I think the comment needs some polishing. What we meant is, first batch is *entirely processed* in memory. I.e., it never gets spilled. As we read the outer tuples and finish joining for the first batch before we start loading the second batch, this works perfectly. However, losing first batch in a rescan scenario is not workable as we will see more outer tuples from later rescans that did not yet join with the first batch, even after we throw out the first batch.

To summarize, if we can guarantee that we no longer need first batch when outer's first batch is joined, we are good to throw (today's scenario). For rescan, we will see new outer tuples (potentially changing set of outer tuples) that need re-joining with first batch. So, we cannot throw out first batch. So, we need to spill first batch now.

File: src/backend/executor/nodeHashjoin.c [hunk 7]
```diff
@@ -888,6 +888,10 @@ ExecHashJoinNewBatch(HashJoinState *hjstate)
 			hashtable->batches[curbatch]->innerside.workfile == NULL))
 
 	{
+		// For rescannable we must complete respilling on first batch
+		if (hjstate->rescannable)
```
Comment (melanieplageman) [inline src/backend/executor/nodeHashjoin.c]: Should we ensure that this is the first pass?

**Comment by melanieplageman** [comment]
During the first pass, if the number of batches is 2, the first inner batch can be loaded into memory, but the  second batch is too large to fit into memory, we need to increase the total batch number to 4, and respill tuples from batch 2 to batch 3 and 4.

For next pass, when we try to load the 1st batch into memory, we use nbatches 4 to select the tuples that we want to load into memory, tuples that don't belong to 1st batch still need to be respilled into other batches.
@foyzur Is this a valid edge case?

**Comment by foyzur** [comment]
@melanieplageman if we double the number of batches during first pass at the time of building hash table (build phase), the successive outer tuple scanning (probe phase) will see the batch number increase and will go through every inner batches, even if there is no outer workfile. That should take care of all respilling and further increase of batches (if we need to respill more during loading of a large inner batch). Therefore, that edge case is already covered. In fact, this edge case was good even before this PR.
